contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,502 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+        }
+
+        // if no final target token is defined, transfer the target token to `transferAddress`
+        if (Token.unwrap(_finalTargetToken) == address(0)) {
+            // safe due to nonreenrant modifier (forwards all available gas if token is native)
+            _targetToken.unsafeTransfer(_transferAddress, sourceAmount);
+            // increment total collected in `transferAddress`
+            _totalCollected += sourceAmount;
+        }
+
+        // if remaining balance is below the min token sale amount, reset the auction
+        if (_amountAvailableForTrading(token) < _minTokenSaleAmounts[token]) {
+            _resetTrading(token, 0);
+        }
+
+        return sourceAmount;
+    }
+
+    function _sellFinalTargetForTarget(uint128 targetAmount) private returns (uint128) {
+        uint128 sourceAmount = expectedTradeInput(_targetToken, targetAmount);
+        // revert if trade requires 0 finalTarget tokens
+        if (sourceAmount == 0) {
+            revert InvalidTrade();
+        }
+
+        // check enough final target token (if final target token is native) has been sent for the trade
+        if (_finalTargetToken == NATIVE_TOKEN) {
+            if (msg.value < sourceAmount) {
+                revert InsufficientNativeTokenSent();
+            } else {
+                // transfer the tokens to the _transferAddress
+                payable(_transferAddress).sendValue(sourceAmount);
+            }
+        } else {
+            // transfer the tokens from the user to the _transferAddress
+            _finalTargetToken.safeTransferFrom(msg.sender, _transferAddress, sourceAmount);
+        }
+
+        // transfer the _targetToken to the user
+        // safe due to nonReentrant modifier (forwards all available gas if native)
+        _targetToken.unsafeTransfer(msg.sender, targetAmount);
+
+        // if final target token is native, refund any excess native token to caller
+        if (_finalTargetToken == NATIVE_TOKEN && msg.value > sourceAmount) {
+            payable(msg.sender).sendValue(msg.value - sourceAmount);
+        }
+
+        // increment total collected in _transferAddress
+        _totalCollected += sourceAmount;
+
+        // update the available target token sale amount
+        _targetTokenSaleAmount.current -= targetAmount;
+
+        // check if remaining target token sale amount is below the min target token sale amount
+        if (_targetTokenSaleAmount.current < _minTokenSaleAmounts[_targetToken]) {
+            // top up the target token sale amount
+            _targetTokenSaleAmount.current = Math
+                .min(_targetToken.balanceOf(address(this)), _targetTokenSaleAmount.initial)
+                .toUint128();
+            // reset the price to price * priceResetMultiplier and restart trading
+            Price memory price = tokenPrice(_targetToken);
+            price.sourceAmount = price.sourceAmount * _priceResetMultiplier;
```

Comment:
Since `_priceResetMultiplier` is set to 2 by default, you are essentially restricting the value of `price.sourceAmount` to 127 bits.

Any value larger than that would cause this transaction to revert on overflow in the expression:
```
price.sourceAmount * _priceResetMultiplier
```
And this might of course become worse if the admin increases the value of `_priceResetMultiplier`.

My suggestion here is therefore to cast one of the operands in that expression to `uint256`.

Reply:
reallistically speaking, the price.sourceAmount for target tokens will be at market price at the point of reset - no one will be trading at a price of 2^127.